File: apps/ios/AgentForge/Packages/AgentForgeCore/Sources/AgentForgeCore/AgentForgeEndpoints.swift
import Foundation

public let defaultAPIBaseURL = "https://agentforge-api-production-5fc1.up.railway.app"
public let defaultDashboardBaseURL = "https://agentforge-web-production.up.railway.app"

public struct AgentForgeEndpoints: Equatable {
    public let apiBaseURL: URL
    public let dashboardBaseURL: URL

    public init(apiBaseURL: URL, dashboardBaseURL: URL) {
        self.apiBaseURL = apiBaseURL
        self.dashboardBaseURL = dashboardBaseURL
    }

    public var healthURL: URL {
        apiBaseURL.appending(path: "health")
    }

    public var readinessURL: URL {
        apiBaseURL.appending(path: "ready")
    }

    public var githubOAuthURL: URL {
        dashboardBaseURL.appending(path: "auth/github/login")
    }
}

public enum EndpointValidationError: LocalizedError, Equatable {
    case missing
    case malformed
    case insecureDeployedURL

    public var errorDescription: String? {
        switch self {
        case .missing:
            "URL is required."
        case .malformed:
            "URL must include an HTTP or HTTPS host."
        case .insecureDeployedURL:
            "Use HTTPS for deployed AgentForge URLs."
        }
    }
}

public enum EndpointValidator {
    public static func normalizedBaseURL(from input: String) throws -> URL {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))

        guard !trimmed.isEmpty else {
            throw EndpointValidationError.missing
        }

        let candidate = trimmed.contains("://") ? trimmed : "https://\(trimmed)"
        guard let components = URLComponents(string: candidate),
              let scheme = components.scheme?.lowercased(),
              let host = components.host?.lowercased(),
              ["https", "http"].contains(scheme)
        else {
            throw EndpointValidationError.malformed
        }

        if scheme == "http", !isLocalDevelopmentHost(host) {
            throw EndpointValidationError.insecureDeployedURL
        }

        guard let url = components.url else {
            throw EndpointValidationError.malformed
        }
        return url
    }

    private static func isLocalDevelopmentHost(_ host: String) -> Bool {
        host == "localhost" || host == "127.0.0.1" || host == "::1" || host.hasSuffix(".local")
    }
}

File: apps/android/app/src/main/java/com/leonardwongly/agentforge/data/AgentForgeEndpoints.kt
package com.leonardwongly.agentforge.data

import java.net.URI

const val DEFAULT_API_BASE_URL = "https://agentforge-api-production-5fc1.up.railway.app"
const val DEFAULT_DASHBOARD_BASE_URL = "https://agentforge-web-production.up.railway.app"

data class AgentForgeEndpoints(
  val apiBaseUrl: String = DEFAULT_API_BASE_URL,
  val dashboardBaseUrl: String = DEFAULT_DASHBOARD_BASE_URL,
) {
  val healthUrl: String = "$apiBaseUrl/health"
  val readinessUrl: String = "$apiBaseUrl/ready"
  val githubOAuthUrl: String = "$dashboardBaseUrl/auth/github/login"
}

object EndpointValidator {
  fun normalizeBaseUrl(input: String): Result<String> {
    val trimmed = input.trim().trimEnd('/')
    if (trimmed.isBlank()) {
      return Result.failure(IllegalArgumentException("URL is required."))
    }

    val withScheme =
      if (trimmed.contains("://")) {
        trimmed
      } else {
        "https://$trimmed"
      }

    val uri =
      runCatching { URI(withScheme) }
        .getOrElse { return Result.failure(IllegalArgumentException("URL is malformed.")) }

    val scheme = uri.scheme?.lowercase()
    val host = uri.host?.lowercase()
    if (scheme !in setOf("https", "http") || host.isNullOrBlank()) {
      return Result.failure(IllegalArgumentException("URL must include an HTTP or HTTPS host."))
    }

    if (scheme == "http" && !host.isLocalDevelopmentHost()) {
      return Result.failure(IllegalArgumentException("Use HTTPS for deployed AgentForge URLs."))
    }

    return Result.success(uri.normalize().toString().trimEnd('/'))
  }

  private fun String.isLocalDevelopmentHost(): Boolean =
    this == "localhost" || this == "127.0.0.1" || this == "10.0.2.2"
}
